### PR TITLE
Added first version of episode 4

### PR DESCRIPTION
**Pull request by dafnevk**
First version of episode 4, using the CIFAR 10 dataset6. See #45 .

### CODE REVIEW

File: _episodes/04-networks-are-like-onions.md [hunk 1]
```diff
@@ -0,0 +1,529 @@
...
+~~~
+{: .output}
+
+> ## Number of parameters
+>
+> Suppose we create a Dense (fully connected) layer with 100 hidden units as input units, how many parameters does this layer have?
+>
+> > ## Solution
+> >
+> > Each of the *dim* inputs is connected with 100 hidden units, and each of the hidden units has a bias. So we have *dim * 100 + 100* weights to learn.
```
Comment (psteinb) [inline _episodes/04-networks-are-like-onions.md:164]: `dim` falls a bit from heaven here. Where does this come from? Also, the solution mixes a lot of technical lingo: 
- hidden units
- bias
- parameters
- weights
- inputs

```suggestion
> > Each entry of the input dimensions, i.e. the `shape` of one signal datum, is connected with 100 neurons of our hidden layer, and each of these neurons has a bias term associated to it. So we have `307300` parameters to learn.
```

File: _episodes/04-networks-are-like-onions.md [hunk 2]
```diff
@@ -0,0 +1,529 @@
...
+> > ~~~
+> > dim * 100 + 100
+> > ~~~
+> > {: .language-python}
```
Comment (psteinb) [inline _episodes/04-networks-are-like-onions.md:168]: I suggest to be a bit more explicit in the solution. The computation needs to pick up the terms so that we avoid confusion.

```suggestion
> > ~~~
> > width, height = (32, 32)
> > n_hidden_neurons = 100
> > n_bias = 100
> > n_input_items = width * height
> > n_parameters = (n_input_items * n_hidden_neurons) + n_bias
> > n_parameters
> > ~~~
> > {: .language-python}
```

File: _episodes/04-networks-are-like-onions.md [hunk 3]
```diff
@@ -0,0 +1,529 @@
...
+    Non-trainable params: 0
+    _________________________________________________________________
+> > ~~~
+> > {: .output}
+> {: .solution}
+{: .challenge}
+
+
+
+Of course, we can decrease the number of units in our hidden layer, but this also decreases the number of patterns our network can remember. Moreover, if we increase the image size, this number of weights will explode, even though the task of recognizing large images is not necessarily more difficult than the task of recognizing small images.
```
Comment (psteinb) [inline _episodes/04-networks-are-like-onions.md:201]:
```suggestion
We can decrease the number of units in our hidden layer, but this also decreases the number of patterns our network can remember. Moreover, if we increase the image size, the number of weights will 'explode', even though the task of recognizing large images is not necessarily more difficult than the task of recognizing small images.
```

File: _episodes/04-networks-are-like-onions.md [hunk 4]
```diff
@@ -0,0 +1,529 @@
...
+
+The solution is that we make the network learn in a 'smart' way. The features that we learn should be similar both for small and large images, and similar features (e.g. edges, corners) can appear anywhere in the image (in mathematical terms: *translation invariant*). We do this by making use of a concepts from image processing that precede Deep Learning.
+
+A **convolution matrix**, or **kernel**, is a matrix transformation that we 'slide' over the image to calculate features at each position of the image. For each pixel, we calculate the matrix product between the kernel and the pixel with its surroundings. A kernel is typically small, between 3x3 and 7x7 pixels. We can for example think of the simple 3x3 kernel:
```
Comment (psteinb) [inline _episodes/04-networks-are-like-onions.md:205]:
```suggestion
A **convolution matrix**, or **kernel**, is a matrix transformation that we 'slide' over the image to calculate features at each position of the image. For each pixel, we calculate the matrix product between the kernel and the pixel with its surroundings. A kernel is typically small, between 3x3 and 7x7 pixels. We can for example think of the 3x3 kernel:
```
'Simple', 'just', 'trivial' and 'straightforward' are terms defined by the author and used by people after years of practice. Thus expert bias is introduced quickly. When reading this, this can discourage learners quickly. I suggest to avoid such adjectives.

File: _episodes/04-networks-are-like-onions.md [hunk 5]
```diff
@@ -0,0 +1,529 @@
...
+~~~
+[[0, 0, 0,],
+ [1, 1, 1],
+ [0, 0, 0]]
+~~~
+{: .language-python}
+This kernel will give a high value to a pixel if it's in the middle of a horizontal edge.
```
Comment (psteinb) [inline _episodes/04-networks-are-like-onions.md:212]: This jumps directly to the conceptional consequence. This will introduce high mental load as for learners it may not be too obvious how the horizontal edge, i.e. high/low values of a fixed intensity related to the kernel. 

I suggest to align the text with this image:
https://github.com/dvgodoy/dl-visuals/blob/main/Convolutions/conv3.png

and then move onto a more explicit example like the one with horizontal edge.

File: _episodes/04-networks-are-like-onions.md [hunk 6]
```diff
@@ -0,0 +1,529 @@
...
+>
+> Suppose we apply a convolutional layer with 100 kernels of size 3 * 3 * 3 (the last dimension applies to the rgb channels) to our images of 32 * 32 * 3 pixels. How many parameters do we have? Assume, for simplicity, that the kernels do not use bias terms. Compare this to the answer of the previous exercise
+>
+> > ## Solution
+> >
+> > We have 100 matrices with 3 * 3 * 3 = 27 values each so that gives 27 * 100 = 2700 weights. This is a magnitude of 100 less than the fully connected layer with 100 units!
+> {: .solution}
+{: .challenge}
+
+
```
Comment (psteinb) [inline _episodes/04-networks-are-like-onions.md:240]: I would cut the lesson here. We should try to focus on the exercises at this point and perhaps "train" an entire network with convolutions for the fun of it. I also end my review here.

**Review by psteinb — CHANGES_REQUESTED** [review]

**Comment by psteinb** [comment]
@dafnevk this is quite some material. I stopped my review after the convolution part. Reading through the remaining material with MaxPooling and Dropout, cutting the lesson after the convolutions appears quite natural to me. 

A subsequent lesson could then add more types of layers for our onion and motivate them. I suggest to work towards merging this in for now and then start a separate PR on a split.

File: _episodes/04-networks-are-like-onions.md [hunk 7]
```diff
@@ -0,0 +1,544 @@
...
+> ## Number of model parameters
+>
+> Suppose we apply a convolutional layer with 100 kernels of size 3 * 3 * 3 (the last dimension applies to the rgb channels) to our images of 32 * 32 * 3 pixels. How many parameters do we have? Assume, for simplicity, that the kernels do not use bias terms. Compare this to the answer of the previous exercise
+>
+> > ## Solution
+> >
+> > We have 100 matrices with 3 * 3 * 3 = 27 values each so that gives 27 * 100 = 2700 weights. This is a magnitude of 100 less than the fully connected layer with 100 units!
+> {: .solution}
+{: .challenge}
+
```
Comment (bpmweel) [inline _episodes/04-networks-are-like-onions.md:254]: I would mention/reitterate something about the performance or expressiveness of convolutional layers or allude to it for later. Because now I had the though: "Ok it's a lot smaller, but that doesn't mean much"

File: _episodes/04-networks-are-like-onions.md (hunk 6 above)
Comment (bpmweel) [reply _episodes/04-networks-are-like-onions.md:255]: I kinda agree that training a NN with convolutions here would be nice.
Introduce a concept and use it, then move on to the next concept.

I do feel that the drop out is very useful to have in this lesson as it helps a lot against overfitting.

File: _episodes/04-networks-are-like-onions.md [hunk 8]
```diff
@@ -0,0 +1,544 @@
...
+sns.lineplot(data=history_df[['accuracy', 'val_accuracy']])
+~~~
+{: .language-python}
+![Output of plotting sample](../fig/04_training_history_1.png)
+~~~
+sns.lineplot(data=history_df[['loss', 'val_loss']])
+~~~
+{: .language-python}
+![Output of plotting sample](../fig/04_training_history_loss_1.png)
+
```
Comment (bpmweel) [inline _episodes/04-networks-are-like-onions.md:330]: This is the first time pandas and seaborn are used in this lesson, so I suggest we add an import statement for them.

**Review by bpmweel — COMMENTED** [review]
I think this is very nice episode, but I have added a few small comments.

I see you're using the active voice a lot, which on the one had is good, but sometimes the sentence does not flow very well with it. For instance: 'We have a look at the CIFAR10 dataset', and 'We take a small sample of the data...'.

File: _episodes/04-networks-are-like-onions.md [hunk 9]
```diff
@@ -0,0 +1,544 @@
...
+import matplotlib.pyplot as plt
+plt.figure(figsize=(10,10))
+for i in range(25):
+    plt.subplot(5,5,i+1)
+    plt.imshow(train_images[i], cmap=plt.cm.binary)
+    plt.axis('off')
+    plt.title(class_names[train_labels[i,0]])
+plt.show()
+~~~
+{: .language-python}
```
[the remark this answers is not in the document]
Comment (dafnevk) [reply _episodes/04-networks-are-like-onions.md:145]: Hm that is strange! This requires some further investigations.
For me, the output pictures match the labels, but depending on the matplotlib version (I think, could also be a different package that causes this) the images are shown upside down.

File: _episodes/04-networks-are-like-onions.md (hunk 6 above)
Comment (dafnevk) [reply _episodes/04-networks-are-like-onions.md:255]: But would you suggest then to not introduce the pooling layers? Or only introduce them after we first train a CNN without pooling layers? 
I think CNNs almost always include some sort of pooling and they are easier to understand than convolutional layers, so I prefer to keep them in the same episode. 

But I will try to add an exercise in between, before we introduce the pooling layer.

**Review by bpmweel — APPROVED** [review]
I think it is good as it is. We can work out other issues later.
